File: internal/csvons/field_helpers.go
package csvons

// requiredFieldValues validates a generated field expression and ensures
// value extraction channel is available. It aborts via failf on invalid input.
func requiredFieldValues(fieldExpr FieldExpr, fieldName string, fields []string, records [][]string) <-chan string {
	if fieldExpr == nil {
		failf("field expression [%s] is nil", fieldName)
		return nil
	}
	vals := fieldExpr.FieldValue(fields, records)
	if vals == nil {
		failf("field expression [%s] cannot resolve values", fieldName)
		return nil
	}
	return vals
}

func requiredFieldOccurrences(fieldExpr FieldExpr, fieldName string, fields []string, records [][]string, ctx ValidationContext) <-chan FieldOccurrence {
	ctx.Field = fieldName
	if fieldExpr == nil {
		failRuntime(ctx, "field expression [%s] is nil", fieldName)
		return nil
	}

	provider, ok := fieldExpr.(fieldOccurrenceProvider)
	if !ok {
		failRuntime(ctx, "field expression [%s] cannot resolve values", fieldName)
		return nil
	}

	occurrences := provider.FieldOccurrences(fields, records)
	if occurrences == nil {
		failRuntime(ctx, "field expression [%s] cannot resolve values", fieldName)
		return nil
	}
	return occurrences
}

File: gui/electron/src/core/field-helpers.ts
import type { FieldExpr } from './field-expr.ts';
import type {
  FieldOccurrence,
  FieldOccurrenceProvider,
} from './field-occurrences.ts';
import { isFieldOccurrenceProvider } from './field-occurrences.ts';
import {
  type ValidationContext,
  makeValidationError,
  type ValidationError,
} from './types.ts';

/**
 * Validates a field expression and returns its values iterable.
 * Returns a ValidationError if the expression is nil or cannot resolve values.
 */
export function requiredFieldValues(
  fieldExpr: FieldExpr | null,
  fieldName: string,
  fields: string[],
  records: string[][],
): Iterable<string> | ValidationError {
  if (!fieldExpr) {
    return makeValidationError({}, 2, `field expression [${fieldName}] is nil`);
  }
  const vals = fieldExpr.fieldValues(fields, records);
  if (!vals) {
    return makeValidationError(
      {},
      2,
      `field expression [${fieldName}] cannot resolve values`,
    );
  }
  return vals;
}

/**
 * Validates a field expression and returns its occurrences iterable.
 * Returns a ValidationError if the expression is nil or cannot resolve values.
 */
export function requiredFieldOccurrences(
  fieldExpr: FieldExpr | null,
  fieldName: string,
  fields: string[],
  records: string[][],
  ctx: ValidationContext,
): Iterable<FieldOccurrence> | ValidationError {
  const fieldCtx = { ...ctx, field: fieldName };

  if (!fieldExpr) {
    return makeValidationError(
      fieldCtx,
      2,
      `field expression [${fieldName}] is nil`,
    );
  }

  if (!isFieldOccurrenceProvider(fieldExpr)) {
    return makeValidationError(
      fieldCtx,
      2,
      `field expression [${fieldName}] cannot resolve values`,
    );
  }

  const occurrences = (fieldExpr as FieldOccurrenceProvider).fieldOccurrences(
    fields,
    records,
  );
  if (!occurrences) {
    return makeValidationError(
      fieldCtx,
      2,
      `field expression [${fieldName}] cannot resolve values`,
    );
  }
  return occurrences;
}

/**
 * Type guard to check if a result is a ValidationError.
 */
export function isValidationError(
  result: unknown,
): result is ValidationError {
  return result instanceof Error && 'code' in result;
}
